File: worker/providers.js
// Provider adapters.
//
// Every provider is described by the same three things, so nothing above this
// file knows which vendor is being talked to:
//
//   models   capability flags the UI needs (does this model take a
//            temperature? does it stream reasoning?)
//   request  turn a normalized chat request into an upstream fetch
//   parse    turn one upstream SSE `data:` payload into normalized events
//
// Normalized events are the only shape the Worker streams to the browser:
//   { type: "text",     text }      assistant output
//   { type: "thinking", text }      reasoning, where the model exposes it
//   { type: "meta",     stopReason, usage: { input, output } }
//   { type: "error",    message }   upstream failure mid-stream
//
// Adding a provider means adding one entry here. Nothing else changes.

// Capability flags per model:
//   temperature  false where the API rejects sampling parameters outright
//   thinking     model streams reasoning we can surface separately
const ANTHROPIC_VERSION = "2023-06-01";

export const PROVIDERS = {
  anthropic: {
    label: "Anthropic",
    keyVar: "ANTHROPIC_API_KEY",
    // Current Anthropic models reject temperature/top_p with a 400, so an
    // unrecognized model is assumed to behave the same way.
    defaultCaps: { temperature: false, thinking: true },
    models: [
      { id: "claude-opus-5", label: "Claude Opus 5", temperature: false, thinking: true },
      { id: "claude-sonnet-5", label: "Claude Sonnet 5", temperature: false, thinking: true },
      { id: "claude-haiku-4-5", label: "Claude Haiku 4.5", temperature: true, thinking: false },
    ],

    request({ key, model, caps, system, messages, temperature, maxTokens }) {
      const body = {
        model,
        max_tokens: maxTokens,
        stream: true,
        messages: messages.map((m) => ({ role: m.role, content: m.content })),
      };
      if (system) body.system = system;
      // Adaptive thinking is on by default on the 5-series; asking for the
      // summary is what makes it visible in the transcript.
      if (caps.thinking) body.thinking = { type: "adaptive", display: "summarized" };
      if (caps.temperature && temperature != null) body.temperature = temperature;

      return {
        url: "https://api.anthropic.com/v1/messages",
        init: {
          method: "POST",
          headers: {
            "x-api-key": key,
            "anthropic-version": ANTHROPIC_VERSION,
            "content-type": "application/json",
          },
          body: JSON.stringify(body),
        },
      };
    },

    parse(data, state) {
      const chunk = JSON.parse(data);
      switch (chunk.type) {
        case "message_start":
          state.input = chunk.message?.usage?.input_tokens ?? 0;
          return [];
        case "content_block_delta":
          if (chunk.delta?.type === "text_delta") return [{ type: "text", text: chunk.delta.text }];
          if (chunk.delta?.type === "thinking_delta")
            return [{ type: "thinking", text: chunk.delta.thinking }];
          return [];
        case "message_delta":
          return [
            {
              type: "meta",
              stopReason: chunk.delta?.stop_reason ?? null,
              usage: { input: state.input ?? 0, output: chunk.usage?.output_tokens ?? 0 },
            },
          ];
        case "error":
          return [{ type: "error", message: chunk.error?.message || "Upstream error" }];
        default:
          return [];
      }
    },
  },

  google: {
    label: "Google Gemini",
    keyVar: "GEMINI_API_KEY",
    defaultCaps: { temperature: true, thinking: false },
    models: [
      { id: "gemini-2.5-pro", label: "Gemini 2.5 Pro", temperature: true, thinking: true },
      { id: "gemini-2.5-flash", label: "Gemini 2.5 Flash", temperature: true, thinking: true },
      {
        id: "gemini-2.5-flash-lite",
        label: "Gemini 2.5 Flash Lite",
        temperature: true,
        thinking: false,
      },
    ],

    request({ key, model, caps, system, messages, temperature, maxTokens }) {
      const body = {
        contents: messages.map((m) => ({
          role: m.role === "assistant" ? "model" : "user",
          parts: [{ text: m.content }],
        })),
        generationConfig: { maxOutputTokens: maxTokens },
      };
      if (system) body.systemInstruction = { parts: [{ text: system }] };
      if (caps.temperature && temperature != null) body.generationConfig.temperature = temperature;

      return {
        url:
          "https://generativelanguage.googleapis.com/v1beta/models/" +
          `${encodeURIComponent(model)}:streamGenerateContent?alt=sse`,
        init: {
          method: "POST",
          headers: { "x-goog-api-key": key, "content-type": "application/json" },
          body: JSON.stringify(body),
        },
      };
    },

    parse(data) {
      const chunk = JSON.parse(data);
      if (chunk.error) return [{ type: "error", message: chunk.error.message || "Upstream error" }];

      const events = [];
      const candidate = chunk.candidates?.[0];
      // Gemini marks reasoning with `thought: true` on the part rather than
      // giving it its own event type.
      for (const part of candidate?.content?.parts ?? []) {
        if (typeof part.text !== "string") continue;
        events.push({ type: part.thought ? "thinking" : "text", text: part.text });
      }
      if (candidate?.finishReason || chunk.usageMetadata) {
        events.push({
          type: "meta",
          stopReason: candidate?.finishReason ?? null,
          usage: {
            input: chunk.usageMetadata?.promptTokenCount ?? 0,
            output: chunk.usageMetadata?.candidatesTokenCount ?? 0,
          },
        });
      }
      return events;
    },
  },

  openai: {
    label: "OpenAI",
    keyVar: "OPENAI_API_KEY",
    defaultCaps: { temperature: true, thinking: false },
    models: [
      { id: "gpt-4.1", label: "GPT-4.1", temperature: true, thinking: false },
      { id: "gpt-4.1-mini", label: "GPT-4.1 mini", temperature: true, thinking: false },
      { id: "gpt-4o", label: "GPT-4o", temperature: true, thinking: false },
      // Reasoning models reject temperature and keep their reasoning private.
      { id: "o3", label: "o3", temperature: false, thinking: false },
      { id: "o4-mini", label: "o4-mini", temperature: false, thinking: false },
    ],

    request({ key, model, caps, system, messages, temperature, maxTokens }) {
      const body = {
        model,
        stream: true,
        stream_options: { include_usage: true },
        max_completion_tokens: maxTokens,
        messages: system ? [{ role: "system", content: system }, ...messages] : messages,
      };
      if (caps.temperature && temperature != null) body.temperature = temperature;

      return {
        url: "https://api.openai.com/v1/chat/completions",
        init: {
          method: "POST",
          headers: { authorization: `Bearer ${key}`, "content-type": "application/json" },
          body: JSON.stringify(body),
        },
      };
    },

    parse(data) {
      if (data === "[DONE]") return [];
      const chunk = JSON.parse(data);
      if (chunk.error) return [{ type: "error", message: chunk.error.message || "Upstream error" }];

      const events = [];
      const choice = chunk.choices?.[0];
      if (choice?.delta?.content) events.push({ type: "text", text: choice.delta.content });
      if (choice?.finish_reason || chunk.usage) {
        events.push({
          type: "meta",
          stopReason: choice?.finish_reason ?? null,
          usage: {
            input: chunk.usage?.prompt_tokens ?? 0,
            output: chunk.usage?.completion_tokens ?? 0,
          },
        });
      }
      return events;
    },
  },
};

// Capability lookup that tolerates a model typed by hand: a model the catalog
// doesn't list still works, it just falls back to the provider's defaults.
export function modelCaps(provider, modelId) {
  const known = provider.models.find((m) => m.id === modelId);
  return known ? { temperature: known.temperature, thinking: known.thinking } : provider.defaultCaps;
}

// What the browser is allowed to know: the catalog, plus whether a key exists.
// Never the key itself.
export function describeProviders(env) {
  return Object.entries(PROVIDERS).map(([id, provider]) => ({
    id,
    label: provider.label,
    configured: Boolean(env[provider.keyVar]),
    keyVar: provider.keyVar,
    models: provider.models,
    defaultCaps: provider.defaultCaps,
  }));
}

File: internal/providers/providers.go
// Package providers adapts each vendor's chat API to one interface.
//
// The Worker version kept per-response parser state in a mutable object passed
// back into every call. Go has a better shape for that: a Provider is
// stateless and hands out a Decoder per response, so the state has an obvious
// owner and lifetime, and two concurrent conversations cannot share it.
package providers

import (
	"fmt"
	"net/http"
	"sort"
)

// Event is the normalized shape everything above this package sees. The JSON
// tags matter: these go to the browser unchanged, and the front end already
// speaks this vocabulary.
type Event struct {
	Type string `json:"type"`

	Text    string         `json:"text,omitempty"`
	ID      string         `json:"id,omitempty"`
	Name    string         `json:"name,omitempty"`
	Input   map[string]any `json:"input,omitempty"`
	Summary string         `json:"summary,omitempty"`
	OK      *bool          `json:"ok,omitempty"`
	Content string         `json:"content,omitempty"`

	StopReason string `json:"stopReason,omitempty"`
	Usage      *Usage `json:"usage,omitempty"`
	Round      *int   `json:"round,omitempty"`
	Message    string `json:"message,omitempty"`
}

type Usage struct {
	Input  int `json:"input"`
	Output int `json:"output"`
}

// Event constructors, so the string literals live in one place.
func TextEvent(text string) Event     { return Event{Type: "text", Text: text} }
func ThinkingEvent(text string) Event { return Event{Type: "thinking", Text: text} }
func ErrorEvent(format string, args ...any) Event {
	return Event{Type: "error", Message: fmt.Sprintf(format, args...)}
}

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// ToolSpec is a tool as the model should see it: vendor-neutral JSON Schema
// that each adapter translates into its own dialect.
type ToolSpec struct {
	Name        string
	Description string
	Schema      map[string]any
}

// ToolResult is what a tool produced, on the way back to the model.
type ToolResult struct {
	ID      string
	Name    string
	OK      bool
	Content string
}

// Caps records where vendors disagree about fundamentals.
type Caps struct {
	// Temperature is false where the API rejects sampling parameters outright
	// rather than ignoring them.
	Temperature bool
	// Thinking is true where the model streams reasoning we can surface.
	Thinking bool
}

type Model struct {
	ID          string `json:"id"`
	Label       string `json:"label"`
	Temperature bool   `json:"temperature"`
	Thinking    bool   `json:"thinking"`
}

// Request is one upstream call.
type Request struct {
	APIKey      string
	Model       string
	Caps        Caps
	System      string
	Messages    []Message
	Temperature *float64
	MaxTokens   int
	Tools       []ToolSpec
	// Extra holds native turns accumulated by the tool loop — assistant turns
	// carrying tool calls, and the results answering them. They are vendor
	// shaped by definition, so they stay opaque above this package.
	Extra []any
}

// Decoder turns one response's SSE payloads into events, and can rebuild the
// turn it just read in the vendor's own shape.
//
// The rebuild is the awkward part of tool calling: the assistant turn holding
// a tool call has to go back verbatim, and reasoning blocks carry signatures
// the vendor validates. So decoders accumulate the native content as it
// streams rather than reconstructing it afterwards.
type Decoder interface {
	Decode(payload []byte) ([]Event, error)
	AssistantTurn() []any
	ResultTurn(results []ToolResult) []any
}

type Provider interface {
	Name() string
	Label() string
	KeyEnv() string
	Models() []Model
	Build(req Request) (*http.Request, error)
	NewDecoder() Decoder
}

// Registry holds the providers that have an API key configured.
type Registry struct {
	byName map[string]Provider
}

func NewRegistry(lookupEnv func(string) string) *Registry {
	registry := &Registry{byName: map[string]Provider{}}
	for _, provider := range []Provider{&Anthropic{}, &Google{}, &OpenAI{}} {
		if lookupEnv(provider.KeyEnv()) != "" {
			registry.byName[provider.Name()] = provider
		}
	}
	return registry
}

func (r *Registry) Get(name string) (Provider, bool) {
	provider, ok := r.byName[name]
	return provider, ok
}

func (r *Registry) Names() []string {
	names := make([]string, 0, len(r.byName))
	for name := range r.byName {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

// CapsFor tolerates a model typed by hand: an unlisted id still works, it just
// inherits the provider's defaults.
func CapsFor(provider Provider, modelID string) Caps {
	for _, model := range provider.Models() {
		if model.ID == modelID {
			return Caps{Temperature: model.Temperature, Thinking: model.Thinking}
		}
	}
	return defaultCaps(provider)
}

func defaultCaps(provider Provider) Caps {
	switch provider.(type) {
	case *Anthropic:
		// Current Anthropic models reject temperature with a 400, so an
		// unrecognized one is assumed to behave the same way.
		return Caps{Temperature: false, Thinking: true}
	default:
		return Caps{Temperature: true, Thinking: false}
	}
}
